src/NServiceBus.Core/Transports/Msmq/QueueCreator.cs

```diff
@@ -39,23 +39,30 @@ void CreateQueueIfNecessary(string address, string identity)
 
             Logger.Debug($"Creating '{address}' if needed.");
 
-            MessageQueue queue;
-            if (MsmqUtilities.TryOpenQueue(msmqAddress, out queue) || MsmqUtilities.TryCreateQueue(msmqAddress, identity, settings.UseTransactionalQueues, out queue))
+            MessageQueue queue = null;
+            try
             {
-                using (queue)
+                if (!MsmqUtilities.TryOpenQueue(msmqAddress, out queue) && MsmqUtilities.TryCreateQueue(msmqAddress, identity, settings.UseTransactionalQueues, out queue))
                 {
-                    Logger.Debug("Setting queue permissions.");
-
-                    try
-                    {
-                        QueuePermissions.SetPermissionsForQueue(queue, identity);
-                    }
-                    catch (MessageQueueException ex)
+                    using (queue)
                     {
-                        Logger.Error($"Unable to set permissions for queue {queue.QueueName}", ex);
+                        Logger.Debug("Setting queue permissions.");
+
+                        try
+                        {
+                            QueuePermissions.SetPermissionsForQueue(queue, identity);
+                        }
+                        catch (MessageQueueException ex)
+                        {
+                            Logger.Error($"Unable to set permissions for queue {queue.QueueName}", ex);
+                        }
                     }
                 }
             }
+            finally
+            {
+                queue?.Dispose();
```

Comment:
This looks vierd, I created a proposal that would clean this up

https://github.com/Particular/NServiceBus/pull/4310

thoughts?